File: pixie/evals/criteria.py
"""Pre-made pass criteria for :func:`~pixie.evals.eval_utils.assert_pass`.

Provides :class:`ScoreThreshold`, a configurable criterion that checks
whether a sufficient fraction of test cases score above a threshold.
"""

from __future__ import annotations

from dataclasses import dataclass

from pixie.evals.evaluation import Evaluation


@dataclass
class ScoreThreshold:
    """Pass criteria: *pct* fraction of inputs must score >= *threshold* on all evaluators.

    Attributes:
        threshold: Minimum score an individual evaluation must reach.
        pct: Fraction of test-case inputs (0.0–1.0) that must pass.
    """

    threshold: float = 0.5
    pct: float = 1.0

    def __call__(
        self,
        results: list[list[Evaluation]],
    ) -> tuple[bool, str]:
        """Evaluate the results matrix and return ``(passed, message)``.

        Args:
            results: Shape ``[inputs][evaluators]``.

        Returns:
            A ``(bool, str)`` tuple: whether the criteria are met and a
            human-readable explanation.
        """
        total_inputs = len(results)
        passing_inputs = 0
        for input_evals in results:
            all_pass = all(e.score >= self.threshold for e in input_evals)
            if all_pass:
                passing_inputs += 1

        if total_inputs > 0 and passing_inputs / total_inputs >= self.pct:
            pct_actual = passing_inputs / total_inputs * 100
            return (
                True,
                f"Pass: {passing_inputs}/{total_inputs} inputs ({pct_actual:.1f}%) "
                f"scored >= {self.threshold} on all evaluators "
                f"(required: {self.pct * 100:.1f}%)",
            )

        pct_best = passing_inputs / total_inputs * 100 if total_inputs > 0 else 0.0
        return (
            False,
            f"Fail: {passing_inputs}/{total_inputs} inputs ({pct_best:.1f}%) "
            f"scored >= {self.threshold} on all evaluators "
            f"(required: {self.pct * 100:.1f}%)",
        )

File: ts/src/evals/criteria.ts
/**
 * Pre-made pass criteria for `assertPass`.
 *
 * Provides `ScoreThreshold`, a configurable criterion that checks
 * whether a sufficient fraction of test cases score above a threshold.
 */

import type { Evaluation } from "./evaluation";

/**
 * Criteria function type: receives results matrix, returns [passed, message].
 */
export type PassCriteria = (
  results: Evaluation[][]
) => [boolean, string];

/**
 * Pass criteria: `pct` fraction of inputs must score >= `threshold`
 * on all evaluators.
 */
export class ScoreThreshold {
  readonly threshold: number;
  readonly pct: number;

  constructor(threshold: number = 0.5, pct: number = 1.0) {
    this.threshold = threshold;
    this.pct = pct;
  }

  /**
   * Evaluate the results matrix and return `[passed, message]`.
   *
   * @param results - Shape `[inputs][evaluators]`.
   */
  __call__(results: Evaluation[][]): [boolean, string] {
    const totalInputs = results.length;
    let passingInputs = 0;
    for (const inputEvals of results) {
      const allPass = inputEvals.every((e) => e.score >= this.threshold);
      if (allPass) passingInputs++;
    }

    if (totalInputs > 0 && passingInputs / totalInputs >= this.pct) {
      const pctActual = (passingInputs / totalInputs) * 100;
      return [
        true,
        `Pass: ${passingInputs}/${totalInputs} inputs (${pctActual.toFixed(1)}%) ` +
          `scored >= ${this.threshold} on all evaluators ` +
          `(required: ${(this.pct * 100).toFixed(1)}%)`,
      ];
    }

    const pctBest =
      totalInputs > 0 ? (passingInputs / totalInputs) * 100 : 0;
    return [
      false,
      `Fail: ${passingInputs}/${totalInputs} inputs (${pctBest.toFixed(1)}%) ` +
        `scored >= ${this.threshold} on all evaluators ` +
        `(required: ${(this.pct * 100).toFixed(1)}%)`,
    ];
  }
}
